### PR TITLE
Feature 5 : Search Todo functionality implemented.

### PR DESCRIPTION
**Pull request by saa-wtag**
[Task-5: Search Todo](https://trello.com/c/4pQs97Xz)
This PR includes Todo Search functionality. Here is an input field with a Search button. User can search the desired todo items in the list by using Todo's Title . If the input box is empty and user click's the Search Button, then whole Todo is being showed.

**Functionalities:**

-  There is a new input field on the top of document with a button along with it.
-  When a search phrase is entered and search button is clicked, relevant todos are shown.

https://github.com/sad-wtag/ToDoApp_VanillaJS/assets/164306406/c7d9cc4f-b9be-4182-b4ba-20dd322f6e0a

Please have a close look on this PR and kindly put your valuable comments.

### CODE REVIEW

File: js/elements.js
Comment (sal-wtag) [inline js/elements.js:1]: @saa-wtag bhai, I think its better to follow `camel case` for element ids.

**Review by sal-wtag — APPROVED** [review]

File: index.html [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+    <div>
+      <input type="text" id="task-input" placeholder="Write your task" />
+      <button id="add-button" disabled>Add Task</button>
+    </div>
+    <div>
+      <h5 id="message" hidden></h5>
+    </div>
+    <div class="tasks-container">
+      <ul id="tasks"></ul>
+    </div>
```
Comment (rza-wtag) [inline index.html]: Are these changes part of this PR.? The changes should reflect only the relevant parts of the specific PR which in this case is the search functionality.

File: js/script.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+const addButtonHandler = () => {
+  const taskTitle = sanitizeInput($taskInput.value);
+  if (taskTitle) {
+    createTask(taskTitle);
+    showToastMessage("Task added successfully!");
+    $addButton.disabled = true;
+  } else {
+    showToastMessage("Please provide a valid title!");
+  }
+};
```
Comment (rza-wtag) [inline js/script.js]: Please make changes that are only related to the specific PR

File: js/script.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    id: new Date().getTime(),
+    title: taskTitle,
+    createdAt: formatDate(new Date()),
+  };
+  tasks.unshift(task);
+  renderTasks(tasks);
+  $taskInput.value = "";
+};
+
+const renderTasks = (tasks = []) => {
```
Comment (rza-wtag) [inline js/script.js:95]: Why are we passing the empty array like this.? Can't it be declared on the global scope.?

File: js/utilities.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    $taskItem.append($inputField, $updateButton, $cancelButton);
+  } else {
+    const $deleteButton = createElement("Delete", "button", () =>
+      deleteTask(task.id)
+    );
+    const $editButton = createElement("Edit", "button", () => editTask(task));
+    const $doneButton = createElement("Done", "button", () =>
+      completeTask(task.id)
+    );
```
Comment (rza-wtag) [inline js/utilities.js]: For better readability create the buttons like this - `const deleteButton = document.createElement("button");` and then you can use addEventListeners accordingly.

Comment (saa-wtag) [reply js/utilities.js]: Here I've created a modular function to handle element creation. So I am passing the element name, element type and the handler function. Then the modular function creates that element for me which helps us to obtain modularity in this project.

File: js/script.js (hunk 3 above)
Comment (saa-wtag) [reply js/script.js:95]: We have declared tasks as global variable but again assigning empty array in the function it helps achieving function Independence, preventing potential runtime errors and increasing code clarity.

File: index.html (hunk 1 above)
Comment (saa-wtag) [reply index.html]: Bhai, I didn't do rebase properly. Now I think you won't see those inconvenience.
Thank you.

File: js/script.js (hunk 2 above)
Comment (saa-wtag) [reply js/script.js]: Bhai, I didn't do rebase properly. Now I think you won't see those inconvenience.
Thank you.

File: js/utilities.js [hunk 5]
```diff
@@ -5,7 +5,7 @@ export const showToastMessage = (message) => {
   $toaster.hidden = false;
   setTimeout(() => {
     $toaster.hidden = true;
-  }, 3000);
+  }, 1500);
```
Comment (rza-wtag) [inline js/utilities.js]: It's better to not make this change to this PR. As it's was primarily done for the Add Tasks PR.

You may use something simple like appending a simple text as an alert, but it should not be DOM blocking, and you can always beautify the toaster in the UI design section.

File: index.html [hunk 6]
```diff
@@ -9,10 +9,17 @@
...
-    <p>Add Tasks</p>
-    <input type="text" id="task-input" placeholder="Write your task" />
-    <button id="add-button" disabled>Add Task</button>
+    <div>
+      <input type="text" id="search-input" placeholder="Search tasks" />
+      <button id="search-button">Search</button>
+    </div>
+    <div>
+      <input type="text" id="task-input" placeholder="Write your task" />
+      <button id="add-button" disabled>Add Task</button>
```
Comment (rza-wtag) [inline index.html:18]: Anything that is not related to this PR should not be in the changes.

Comment (saa-wtag) [reply index.html:18]: Bhai, I have added search related things here and for convenience added a div to add task element, which wasn't required earlier.

File: js/utilities.js (hunk 5 above)
Comment (saa-wtag) [reply js/utilities.js]: okay bhai

**Review by rza-wtag — APPROVED** [review]

File: .vscode/settings.json [hunk 7]
```diff
@@ -1,3 +0,0 @@
-{
```
Comment (maa-wtag) [inline .vscode/settings.json]: please add these .vscode file into .gitignore, these are your local files and not related to the project, they are not supposed to be pushed.

File: js/script.js [hunk 8]
```diff
@@ -19,22 +25,40 @@ const addButtonHandler = () => {
...
+      renderTasks(filteredTasks);
+    } else {
+      showToastMessage("No tasks found matching the search.");
+    }
+  } else {
+    renderTasks(tasks);
+  }
+  $searchInput.value = "";
+};
+
```
Comment (maa-wtag) [inline js/script.js]: we can make this more readable and clean

```suggestion
const handleSearchTasks = () => {
  // Sanitize and retrieve the input value
  const searchTitle = sanitizeInput($searchInput.value.trim());

  // Clear the input field
  $searchInput.value = "";

  // Use early return to handle empty search scenario
  if (!searchTitle) {
    renderTasks(tasks);
    return;
  }

  // Filter tasks based on the sanitized input
  const filteredTasks = tasks.filter(task => 
    task.title.toLowerCase().includes(searchTitle.toLowerCase())
  );

  // Render based on the filtering result
  if (filteredTasks.length > 0) {
    renderTasks(filteredTasks);
  } else {
    showToastMessage("No tasks found matching the search.");
  }
};


```

File: js/script.js [hunk 9]
```diff
@@ -19,22 +25,40 @@ const addButtonHandler = () => {
...
+  const searchTitle = sanitizeInput($searchInput.value);
+
+  if (searchTitle) {
+    const filteredTasks = tasks.filter((task) =>
+      task.title.toLowerCase().includes(searchTitle.toLowerCase())
+    );
+    if (filteredTasks.length > 0) {
+      renderTasks(filteredTasks);
+    } else {
+      showToastMessage("No tasks found matching the search.");
```
Comment (maa-wtag) [inline js/script.js]: could we keep all the messages in a utility file as constants and use it from there?

**Review by maa-wtag — CHANGES_REQUESTED** [review]

File: js/script.js (hunk 9 above)
Comment (saa-wtag) [reply js/script.js]: Done, bhai.

File: js/script.js (hunk 8 above)
Comment (saa-wtag) [reply js/script.js]: Done, bhai.

File: .vscode/settings.json (hunk 7 above)
Comment (saa-wtag) [reply .vscode/settings.json]: resolved.

**Review by maa-wtag — APPROVED** [review]
The vscode file changes are still showing in the PR. please take care of it before passing to the next reviewer.

File: js/script.js [hunk 10]
```diff
@@ -12,29 +21,51 @@ const addButtonHandler = () => {
...
+    return;
+  }
+  // Filter tasks based on the sanitized input
+  const filteredTasks = tasks.filter((task) =>
+    task.title.toLowerCase().includes(searchTitle.toLowerCase())
+  );
+  // Render based on the filtering result
+  if (filteredTasks.length > 0) {
+    renderTasks(filteredTasks);
+  } else {
```
Comment (hhu-wtag) [inline js/script.js:48]: Although adding comments in code might help now and then, adding comments too much/unnecessarily will produce what is known as "code smell". This is a general anti pattern that is often not advised.

Only add comment where it is absolutely necessary. Otherwise we should write self documented code. This is one of the reasons why we strive to write simpler, modular code and avoid complex oneliners which might impress myself(the developer) and help me have a good nights sleep but will make me question the same code 5 months later.
:p

**Review by hhu-wtag — APPROVED** [review]
